### PR TITLE
Fixed camera stacking output properties

### PR DESCRIPTION
**Pull request by lukaschod**
### Purpose of this PR
Base camera output properties editor changes are now also applied to all cameras in stack.
This is needed for Canvas to correctly work if any overlay camera output settings are out of sync with base camera.
It happens, because our core code still depends on these properties.
https://fogbugz.unity3d.com/f/cases/1288105/
https://fogbugz.unity3d.com/f/cases/1311268/
https://fogbugz.unity3d.com/f/cases/1252265/
https://fogbugz.unity3d.com/f/cases/1291872/

Replace context.SetupCameraProperties(camera) with URP shader property setup.
https://fogbugz.unity3d.com/f/cases/1226940/
This includes new methods:
- SetPerCameraClippingPlaneProperties(cmd, in cameraData);
- SetPerCameraBillboardProperties(cmd, ref cameraData);
- Projection setting in SetPerCameraShaderVariables.

In prefab changes for camera stack are not correctly showed in GUI.
Also added header as without it, it was resulting in 1 pixel size prefab area. Also I think it does not look that bad with header.
https://fogbugz.unity3d.com/f/cases/1308717/
Before
image
After
image

Changed the layout of error icon with camera render type. With @TheoWong-pixel we decided that it looks a bit better visually, more consistant with new warning and also it does not overflow text that much when shrunk.
Also changed error icon to be consistent with helpbox, which account unity skin, point per pixel image change. More over addded fix me buttom.
Before
resize-stack
After
image
image

Added new warning in camera stack. To warn user when one of the overlay camera has output properties out of sync with base camera.
image

Errors and warnings updated after removing camera that had problems.
Before
remove-issue

Info boxes are now in same indent as camera stack list.

Roughly fixes 52 user pain.
---
### Testing status
- [ ] Local gfx tests
- [ ] Test is overlay camera rendering did not regressed in any way (Especially if these properties did not regressed https://docs.unity3d.com/Manual/SL-UnityShaderVariables.html)
- [ ] Test if UI changes are good user experience.
- [ ] Urp QA manaul pass.
- [ ] XR QA manual pass.
- [ ] Mobile QA manual pass.

To make testing easier I made script that automatically compares if any property regressed.
image
image
Basically it quaeries all these properties before changes and after changes and checks if there is any difference. Also checks if all values are not zero.
I will keep this script until QA will finish, maybe I will find a way to create legit test from it.

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.universal/Runtime/TestCameraProperties.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
+using UnityEngine;
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Runtime/TestCameraProperties.cs:1]: This is only for QA testing

File: com.unity.render-pipelines.universal/Runtime/ShowCameraProperties.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+[ExecuteAlways]
+public class ShowCameraProperties : MonoBehaviour
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Runtime/ShowCameraProperties.cs:6]: This is only for QA testing

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 3]
```diff
@@ -207,6 +209,104 @@ void SetPerCameraShaderVariables(CommandBuffer cmd, ref CameraData cameraData)
             cmd.SetGlobalVector(ShaderPropertyId.scaledScreenParams, new Vector4(scaledCameraWidth, scaledCameraHeight, 1.0f + 1.0f / scaledCameraWidth, 1.0f + 1.0f / scaledCameraHeight));
             cmd.SetGlobalVector(ShaderPropertyId.zBufferParams, zBufferParams);
             cmd.SetGlobalVector(ShaderPropertyId.orthoParams, orthoParams);
+
+            // TODO: Find out if this is needed
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:213]: Need someone to confirm is ths needed for SRP in general?

Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:213]: We do not use this in URP, but maybe we want to keep it if users depend on it?

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 4]
```diff
@@ -523,8 +623,58 @@ public void Execute(ScriptableRenderContext context, ref RenderingData rendering
...
+                        TestCameraProperties.SampleCameraPropertiesOld(cmd);
+
+                        // Set new properties
+                        SetPerCameraShaderVariables(cmd, ref cameraData); // TODO: Remove after the testing
+                        SetCameraMatrices(cmd, ref cameraData, true);
+                        SetPerCameraClippingPlaneProperties(cmd, in cameraData); // TODO: Move this is to SetCameraMatrices, once we get rid of context.SetupCameraProperties
+                        SetPerCameraBillboardProperties(cmd, ref cameraData);
+
+                        // TODO: Find out if this is needed
+                        /*#if GFX_SUPPORTS_SINGLE_PASS_STEREO
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:647]: Need someone from XR team to cofirm is this needed?

File: com.unity.render-pipelines.universal/Shaders/FetchCameraPropertiesShader.shader [hunk 5]
```diff
@@ -0,0 +1,70 @@
+Shader "Hidden/FetchCameraPropertiesShader"
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Shaders/FetchCameraPropertiesShader.shader:1]: This is only for QA testing

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 6]
```diff
@@ -207,6 +209,104 @@ void SetPerCameraShaderVariables(CommandBuffer cmd, ref CameraData cameraData)
...
+            float c = worldRight.x * billboardTangent.x + worldRight.z * billboardTangent.z;
+            cameraXZAngle = Mathf.Atan2(s, c);
+
+            // convert to [0,2PI)
+            if (cameraXZAngle < 0)
+                cameraXZAngle += 2 * Mathf.PI;
+        }
+
+        // TODO: Move this to SetPerCameraShaderVariables once we get rid of context.SetupCameraProperties
+        private void SetPerCameraClippingPlaneProperties(CommandBuffer cmd, in CameraData cameraData)
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:288]: @phi-lira maybe we should move it in this PR? WDYT?
The reason I kept it separate to stick to original plan, where base camera setup will not be changed.

File: com.unity.render-pipelines.core/Editor/CoreEditorUtils.cs [hunk 7]
```diff
@@ -77,7 +77,17 @@ static CoreEditorUtils()
...
-            EditorGUILayout.HelpBox(text, MessageType.Warning);
+            DrawFixMeBox(text, MessageType.Warning, action);
+        }
+
+        // UI Helpers
+        /// <summary>Draw a Fix button</summary>
+        /// <param name="text">Displayed message</param>
+        /// <param name="action">Action performed when fix buttom is clicked</param>
+        /// <param name="messageType">Type of the message</param>
+        public static void DrawFixMeBox(string text, MessageType messageType, Action action)
```
Comment (lukaschod) [inline com.unity.render-pipelines.core/Editor/CoreEditorUtils.cs:88]: It is strange that nobody ever needed error fixme box. Maybe I am violating some concept? @TheoWong-pixel WDYT?

File: com.unity.render-pipelines.universal/Runtime/TestCameraProperties.cs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+using Unity.Mathematics;
+using UnityEngine.Rendering;
+
+// TODO: Remove after the testing
+internal static class TestCameraProperties
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Runtime/TestCameraProperties.cs:6]: This is only for QA testing

Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/TestCameraProperties.cs:6]: Yea this is such a hack :D, but suprisingly it works

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 9]
```diff
@@ -207,6 +209,104 @@ void SetPerCameraShaderVariables(CommandBuffer cmd, ref CameraData cameraData)
...
+        {
+            // We need this to correctly handle projection inversion if needed
+            Matrix4x4 projectionMatrix = GL.GetGPUProjectionMatrix(cameraData.GetProjectionMatrix(), cameraData.targetTexture != null);
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:291]: Shoudn't it be flipped only if platform uv starts at top? We also cache gpu projection matrix in cmaeraData.

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 6 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:288]: That's fine to move in follow up PR if we don't want to introduce risk to base camera changes in this one.

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 10]
```diff
@@ -523,8 +623,58 @@ public void Execute(ScriptableRenderContext context, ref RenderingData rendering
...
+                        SetCameraMatrices(cmd, ref cameraData, true);
+
+                        // TODO: Remove after the testing
+                        TestCameraProperties.SampleCameraPropertiesOld(cmd);
+
+                        // Set new properties
+                        SetPerCameraShaderVariables(cmd, ref cameraData); // TODO: Remove after the testing
+                        SetCameraMatrices(cmd, ref cameraData, true);
+                        SetPerCameraClippingPlaneProperties(cmd, in cameraData); // TODO: Move this is to SetCameraMatrices, once we get rid of context.SetupCameraProperties
+                        SetPerCameraBillboardProperties(cmd, ref cameraData);
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:644]: It would be good to add a runtime test to make sure that the constants that we setup are the same as the SetupCameraProperties.

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 11]
```diff
@@ -207,6 +209,104 @@ void SetPerCameraShaderVariables(CommandBuffer cmd, ref CameraData cameraData)
...
+            {
+                return new Vector4(plane.normal.x, plane.normal.y, plane.normal.z, plane.distance);
+            }
+
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes0, PlaneToVector(planes[0]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes1, PlaneToVector(planes[1]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes2, PlaneToVector(planes[2]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes3, PlaneToVector(planes[3]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes4, PlaneToVector(planes[4]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes5, PlaneToVector(planes[5]));
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:309]: It's weird to me we upload 6 individual constants instead of a vector array, is that due to built-in compatibility?

File: com.unity.render-pipelines.universal/Runtime/TestCameraProperties.cs [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+            }
+        }
+        cmd.DrawMesh(SomeMesh, Matrix4x4.Translate(Vector3.zero), FetchCameraPropertiesMaterialOld);
+
+        CameraPropertiesData[] array = new CameraPropertiesData[1];
+        CameraPropertiesDataBufferOld.GetData(array);
+        DataOld = array[0];
+    }
+
+    public static void SampleCameraPropertiesNew(CommandBuffer cmd)
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/TestCameraProperties.cs:57]: Can't we merge these two functions with some arguments? Makes it easier to maintain imo.

**Review by phi-lira — APPROVED** [review]
LGTM. Minor comments.

File: com.unity.render-pipelines.universal/Runtime/TestCameraProperties.cs (hunk 12 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/TestCameraProperties.cs:57]: Yeap, but this is test script, which currently wont be landed to master.

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 11 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:309]: This is from builtin, I think I can make it array, going to check

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 10 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:644]: Well the problem is that fetching setted properties is quite tricky. My current solution is only for manual testing as it is super hacky.

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 9 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:291]: noice fixed

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 13]
```diff
@@ -207,6 +209,104 @@ void SetPerCameraShaderVariables(CommandBuffer cmd, ref CameraData cameraData)
...
+
+        /// <summary>
+        /// Set camera billboard properties.
+        /// </summary>
+        /// <param name="cmd">CommandBuffer to submit data to GPU.</param>
+        /// <param name="cameraData">CameraData containing camera matrices information.</param>
+        void SetPerCameraBillboardProperties(CommandBuffer cmd, ref CameraData cameraData)
+        {
+            Matrix4x4 worldToCameraMatrix = cameraData.GetViewMatrix();
+            Vector3 cameraPos = cameraData.camera.transform.position;
```
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:234]: Fixed

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 10 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:642]: Isn't it something that we can at least make sure to run on playmode in one core platform?

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 14]
```diff
@@ -193,20 +193,118 @@ void SetPerCameraShaderVariables(CommandBuffer cmd, ref CameraData cameraData)
...
+        private static void CalculateBillboardProperties(
+            in Matrix4x4 worldToCameraMatrix,
+            out Vector3 billboardTangent,
+            out Vector3 billboardNormal,
+            out float cameraXZAngle)
+        {
+            Matrix4x4 cameraToWorldMatrix = worldToCameraMatrix;
+            cameraToWorldMatrix = cameraToWorldMatrix.transpose;
+
+            // TODO
```
Comment (martint-unity) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:257]: remove todo?

**Review by martint-unity — APPROVED** [review]

**Review by TheoWong-pixel — APPROVED** [review]

**Review by thomas-zeng — APPROVED** [review]

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 4 above)
Comment (thomas-zeng) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:645]: I don't think they are needed anymore.

**Comment by thomas-zeng** [comment]
Adding XR QA Ryan as reviewer.
Suggest to test camera stacking in URP XR and check for regressions.

**Comment by catercousin** [comment]
@Appreciate how awesome and tidy this PR description is 🤩

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs [hunk 15]
```diff
@@ -251,44 +251,63 @@ void DrawElementCallback(Rect rect, int index, bool isActive, bool isFocused)
...
+                    typeError ?
+                    EditorGUIUtility.TrTextContent(type.GetName(), "Not a supported type", m_ErrorIcon) :
+                    EditorGUIUtility.TrTextContent(type.GetName());
+
+                EditorGUI.BeginProperty(rect, GUIContent.none, element);
+                var labelWidth = EditorGUIUtility.labelWidth;
+                EditorGUIUtility.labelWidth -= 20f;
+
+                Vector2 cachedIconSize = EditorGUIUtility.GetIconSize();
+                EditorGUIUtility.SetIconSize(new Vector2(rect.height, rect.height)); // To prevent icon shrinking
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:298]: minor:You have the scope `new EditorGUIUtility.IconSizeScope`

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs [hunk 16]
```diff
@@ -499,23 +666,39 @@ void DrawStackSettings()
...
                 m_SerializedCamera.serializedObject.ApplyModifiedProperties();
 
-                if (errorCameras.Any())
+                EditorGUI.indentLevel--;
+                if (m_TypeErrorCameras.Any())
                 {
-                    string errorString = "These cameras are not of a valid type:\n";
-                    string validCameras = "";
-                    foreach (var errorCamera in errorCameras)
+                    string message = "";
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:672]: Please use a `StringBuilder`

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs [hunk 17]
```diff
@@ -608,6 +791,33 @@ private bool IsAnyRendererHasPostProcessingEnabled(UniversalRenderPipelineAsset
             return false;
         }
 
+        private static Texture2D LoadConsoleIcon(bool isError)
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:794]: I think this method should be moved to `CoreEditorUtils` and `m_ErrorIcon` and `m_WarningIcon` to `CoreEditorStyles`

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs [hunk 18]
```diff
@@ -499,23 +666,39 @@ void DrawStackSettings()
                 m_LayerList.DoLayoutList();
                 m_SerializedCamera.serializedObject.ApplyModifiedProperties();
 
-                if (errorCameras.Any())
+                EditorGUI.indentLevel--;
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:669]: I think we shouldn't indent up the helpboxes, I think you should indent the list. @TheoWong-pixel 

image

You can use this to align the list properly:


```
var rect = GUILayoutUtility.GetRect(1, m_LayerList.GetHeight(), GUILayout.ExpandWidth(true));
m_LayerList.DoList(EditorGUI.IndentedRect(rect));
```

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs [hunk 19]
```diff
@@ -463,6 +488,148 @@ public override void OnInspectorGUI()
             m_SerializedCamera.serializedObject.ApplyModifiedProperties();
         }
 
+        private void UpdateStackCemerasToOverlay()
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:491]: question: I wonder if it makes sense to handle a `List<CameraRenderType> validCameraTypes`  when then, here we are assuming that the only valid type is `Overlay`

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs [hunk 20]
```diff
@@ -463,6 +488,148 @@ public override void OnInspectorGUI()
...
+        {
+            int cameraCount = m_SerializedCamera.cameras.arraySize;
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:493]: minor: I would add  couple of methods to `UniversalRenderPipelineSerializedCamera`

```
        /// <summary>
        /// Returns a <see cref="Camera"/> at a given index
        /// </summary>
        /// <param name="index">The index to fetch the camera</param>
        /// <returns><see cref="Camera"/> if the index is valid, otherwise returns null</returns>
        public Camera this[int index]
        {
            get
            {
                if (index < 0 || index >= stackCamerasCount)
                    return null;

                // Return the camera on that index
                return m_SerializedCamera.cameras
                    .GetArrayElementAtIndex(index)
                    .objectReferenceValue as Camera;
            }
        }

        /// <summary>
        /// Returns the number of cameras on the stack
        /// </summary>
        public int stackCamerasCount => cameras.arraySize;
```

The code will be self contained, and you can use `UpdateStackCemerasToOverlay()` and on `UpdateStackCamerasOutput()`

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs [hunk 21]
```diff
@@ -463,6 +488,148 @@ public override void OnInspectorGUI()
...
+                SerializedProperty cameraProperty = m_SerializedCamera.cameras.GetArrayElementAtIndex(i);
+                Camera overlayCamera = cameraProperty.objectReferenceValue as Camera;
+                if (overlayCamera != null)
+                    UpdateStackCameraOutput(overlayCamera);
+            }
+        }
+
+        private void UpdateStackCameraOutput(Camera camera)
+        {
+            Undo.RecordObject(camera, "Inspector Overlay Camera");
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:529]: Undo stack must be localized

https://confluence.unity3d.com/display/LOC/Editor+Localization+Coding+Guidelines

**Review by alex-vazquez — CHANGES_REQUESTED** [review]

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs (hunk 15 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:298]: noice, fixed

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs (hunk 16 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:672]: Updated. Hoever in theory we do not win much there is a good article https://michaelscodingspot.com/challenging-the-c-stringbuilder/. As most of the time it is gona be 3-4 concetations and we still pay the price of allocation from string builder.

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs (hunk 21 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:529]: Fixed, I checked the urp/core code seems like we do not do this, maybe we need jira for that?

Comment (alex-vazquez) [reply com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:529]: Yeah, a JIRA would be nice

**Review by alex-vazquez — APPROVED** [review]

**Review by erikabar — APPROVED** [review]
No PR issues found, testing doc: https://docs.google.com/document/d/1tfZ0a7A6vlv1NNOqimph-ZNfj94put9Izg2yc8CU8Z0/edit#

**Review by ryanhy-unity — APPROVED** [review]
Did some XR checks with the xrautomated URP end to end test which has quite a few graphics features in it including a UI overlay camera stack.  Everything looked good in the following configurations:
Mock HMD on Windows
Rift on Windows (DX11)
Quest (android) using GLES3 and Vulkan

**Review by oleks-k — APPROVED** [review]
Looks good (with the editions)!
